Allow responding asynchronously to OnionMessage

shaavan opened the pull request:
This PR addresses second part of #2882 and builds on #2907

This PR converts `handle_onion_message_resopnse` into the public function which can be used for asynchronous response to a separately created ResponseInstruction.
Also, this introduces a new variant of ResponseInstruction. This allows creating a reply_path for the receiver while responding to their message.

shaavan commented:
Updated from [pr2996.01](https://github.com/shaavan/rust-lightning/commits/pr2996.01) to [pr2996.02](https://github.com/shaavan/rust-lightning/commits/pr2996.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.01..pr2996.02)):

Update:
1. Synced with base PR (#2907) to solve merge conflicts.

shaavan commented:
Updated from [pr2996.02](https://github.com/shaavan/rust-lightning/commits/pr2996.02) to [pr2996.03](https://github.com/shaavan/rust-lightning/commits/pr2996.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.02..pr2996.03)):
Addressed @jkczyz comments:

Updates:
1. Address nit comments.
2. Expanded return type of `handle_onion_message_response`. This change allows access to the success/failure status of sending the response to the user which they can handle appropriately.

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -977,7 +977,9 @@ where
 
-	fn handle_onion_message_response<T: OnionMessageContents>(
+	/// Handles the response to an [`OnionMessage`] based on it's [`ResponseInstruction`]
+	/// and enqueueing it for further processing.
```
Thanks for the suggestions!
Addressed in [989d8a6](https://github.com/lightningdevkit/rust-lightning/commit/989d8a683433390b4ae969cf3220931f61461c2f)

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -977,7 +977,9 @@ where
+	/// and enqueueing it for further processing.
+	pub fn handle_onion_message_response<T: OnionMessageContents>(
 		&self, response: ResponseInstruction<T>
 	) {
```
After our discussion offline. I have updated the return type to Result<Option<SendSuccess>, SendError> in [989d8a6](https://github.com/lightningdevkit/rust-lightning/commit/989d8a683433390b4ae969cf3220931f61461c2f)
Thanks for the pointers!

shaavan commented:
Updated from [pr2996.03](https://github.com/shaavan/rust-lightning/commits/pr2996.03) to [pr2996.04](https://github.com/shaavan/rust-lightning/commits/pr2996.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.03..pr2996.04)):

Updates:
1. Rebase on main, to resolve conflicts post #2907 merge.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -273,6 +273,15 @@ impl Responder {
 		})
 	}
+
+	/// Creates the appropriate [`ResponseInstruction`] for a given response.
```
Yeah, I think you can at least specify the exact `ResponseInstruction` variant for each, which describe the difference.

shaavan replied:
Updated the docs in [2d769d6](https://github.com/lightningdevkit/rust-lightning/commit/2d769d683b2ae9de8d04c23c6bf24684b902eac4)
Let me know if they seem sound now!
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -977,19 +1007,30 @@ where
+		};
+
+		let message_type = response.message.msg_type();
+		let res = self.find_path_and_enqueue_onion_message(
+			response.message, Destination::BlindedPath(response.reply_path), reply_path,
+			format_args!(
+				"when responding to {} onion message with path_id {:02x?}",
+				message_type,
+				response.path_id
+			)
+		);
+
+		match res {
+			Ok(result) => Ok(Some(result)),
+			Err(result) => Err(result)
 		}
```
Use `Option::map` for brevity:

```rust
self.find_path_and_enqueue_onion_message(
    // ...
).map(|result| Some(result))
```

shaavan replied:
Done! [2d769d6](https://github.com/lightningdevkit/rust-lightning/commit/2d769d683b2ae9de8d04c23c6bf24684b902eac4)
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -977,19 +1007,30 @@ where
-			);
+	) -> Result<Option<SendSuccess>, SendError> {
+		let (response, reply_path) = match response {
+			ResponseInstruction::WithReplyPath(response) => (response, self.create_blinded_path().ok()),
```
Should we return an error if `create_blinded_path` fails? Otherwise, we'd send an onion message without a reply path.

shaavan replied:
That sounds like a good idea! Updated in [2d769d6](https://github.com/lightningdevkit/rust-lightning/commit/2d769d683b2ae9de8d04c23c6bf24684b902eac4)
Thanks for the suggestion!

shaavan commented:
Updated from [pr2996.04](https://github.com/shaavan/rust-lightning/commits/pr2996.04) to [pr2996.05](https://github.com/shaavan/rust-lightning/commits/pr2996.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.04..pr2996.05)):
Addressed @TheBlueMatt, @jkczyz comments

Updates:
1. Expanded `respond()`, and `respond_with_reply_path()` documentation.
2. Updated test to check if we got the expected variant of SendSuccess, when calling `handle_onion_message_response`.
3. Refactor the code to improve brevity.
4. Update handle_onion_message response to fail in case we fail to create a blinded_path.
5. Fix the Fuzz test.

shaavan commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -326,6 +326,37 @@ fn three_blinded_hops() {
+	let response_instruction = nodes[0].custom_message_handler.handle_custom_message(message, responder);
+
+	// 6. Simulate Alice asynchronously responding back to Bob with a response.
+	let _ = nodes[0].messenger.handle_onion_message_response(response_instruction);
```
Updated the test in [2d769d6](https://github.com/lightningdevkit/rust-lightning/commit/2d769d683b2ae9de8d04c23c6bf24684b902eac4)
Now it shall pass only when we get the expected variant of result.
Thanks for the suggestion!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -265,14 +265,31 @@ impl Responder {
+	/// Creates a [`ResponseInstruction::WithoutReplyPath`] variant of [`ResponseInstruction`]
+	/// for a given response.
+	///
+	/// [`ResponseInstruction::WithoutReplyPath`] indicates that the response should be sent
+	/// without creating a reply path for the receiver to use to send back a response to us.
```
This part isn't necessary as it simply repeats the linked docs. Generally, a short summary may be fine, but we should avoid duplicating the docs otherwise they risk getting out of sync.

shaavan replied:
That makes sense!
Addressed in [8d5f9dc](https://github.com/lightningdevkit/rust-lightning/commit/8d5f9dce49b9b2b49a2e67d52a3021325ac3fcb3)
Thanks a lot for the suggestion

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -265,14 +265,31 @@ impl Responder {
 	}
 
-	/// Creates the appropriate [`ResponseInstruction`] for a given response.
+	/// Creates a [`ResponseInstruction::WithoutReplyPath`] variant of [`ResponseInstruction`]
+	/// for a given response.
```
No need to repeat the enum name. Let's aim to be less verbose by conveying the most information in fewer words.

```rust
/// Creates a [`ResponseInstruction::WithoutReplyPath`] for sending the given `response`.
```

shaavan replied:
Done! [8d5f9dc](https://github.com/lightningdevkit/rust-lightning/commit/8d5f9dce49b9b2b49a2e67d52a3021325ac3fcb3)
Thanks!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -351,7 +351,13 @@ fn async_response_over_one_blinded_hop() {
+
+	match res {
+		Ok(Some(SendSuccess::Buffered)) => assert!(true),
+		_ => panic!("Unexpected event"),
```
Event?

shaavan replied:
Oops!
I meant to say something about the unexpected "state" of the result, but I confused it with the word "event"! 😅
Thanks for catching it!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -351,7 +351,13 @@ fn async_response_over_one_blinded_hop() {
 
 	// 6. Simulate Alice asynchronously responding back to Bob with a response.
-	nodes[0].messenger.handle_onion_message_response(response_instruction);
+	let res = nodes[0].messenger.handle_onion_message_response(response_instruction);
+
+	match res {
+		Ok(Some(SendSuccess::Buffered)) => assert!(true),
+		_ => panic!("Unexpected event"),
+	}
```
It's a bit more idiomatic to `assert_eq`:

```rust
assert_eq!(
	nodes[0].messenger.handle_onion_message_response(response_instruction),
	Ok(Some(SendSuccess::Buffered)),
);

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1017,24 +1017,24 @@ where
 	/// Handles the response to an [`OnionMessage`] based on its [`ResponseInstruction`],
 	/// enqueueing any response for sending.
-	pub fn handle_onion_message_response<T: OnionMessageContents>(
+	pub fn handle_onion_message_response<T: OnionMessageContents> (
```
Remove unnecessary space.

shaavan replied:
Done! [8d5f9dc](https://github.com/lightningdevkit/rust-lightning/commit/8d5f9dce49b9b2b49a2e67d52a3021325ac3fcb3)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -926,6 +946,24 @@ where
+		self.message_router
+			.create_blinded_paths(recipient, peers, secp_ctx)
+			.and_then(|paths| paths.into_iter().next().ok_or(()))
+			.map_err(|_| SendError::PathNotFound)
```
Seems like when this is used with `handle_onion_message_response`, `SendError::PathNotFound` may returned in two different scenarios:
- when creating a reply path (here)
- when finding a path to send the response

The docs on `SendError::PathNotFound` indicate the latter scenario. I'm not sure if differentiating the two is that important, though, since in the latter case we log. Maybe we should log in both cases?

shaavan replied:
I think logging for the failed `reply_path` case would be a reasonable thing to do!
Introduce a logging mechanism in the `handle_onion_message_response` for it in [8d5f9dc](https://github.com/lightningdevkit/rust-lightning/commit/8d5f9dce49b9b2b49a2e67d52a3021325ac3fcb3)
Let me know if that looks good!
Thanks for the suggestion!

jkczyz replied:
Let's also update the docs on `SendError::PathNotFound`.

shaavan replied:
Updated the `SendError::PathNotFound` docs in [27899b1](https://github.com/lightningdevkit/rust-lightning/commit/27899b1a04bfd78db34f2c22ad489f1a6fd4e5e4)
Let me know if they look apt!
Thanks for the suggestion!

shaavan commented:
Updated from [pr2996.05](https://github.com/shaavan/rust-lightning/commits/pr2996.05) to [pr2996.06](https://github.com/shaavan/rust-lightning/commits/pr2996.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.05..pr2996.06)):
Addressed @jkczyz comments

Updates:
1. Simplify the respond, and respond_with_reply_path docs to make them concise.
2. Simplify the result check in the added test.
3. Introduced logging in handle_onion_message_response if create_blinded_path fails to create reply_path.

shaavan commented:
Updated from [pr2996.06](https://github.com/shaavan/rust-lightning/commits/pr2996.06) to [pr2996.07](https://github.com/shaavan/rust-lightning/commits/pr2996.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.06..pr2996.07)):

Updates:
1. Rebase on main, to fix ci failures.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -260,21 +260,30 @@ pub struct Responder {
 	}
 
-	/// Creates the appropriate [`ResponseInstruction`] for a given response.
+	/// Creates a [`ResponseInstruction::WithoutReplyPath`] for a given response.
```
How about we add a second sentence? Just want to avoid repeating the linked docs.

```rust
/// Use when the recipient doesn't need to send back a reply.
```

Similarly for the other method.

shaavan replied:
I think this line will aptly explain the purpose of this function without introducing a risk of future docs desync.
Thanks for the suggestion!
Updated docs in [57bca64](https://github.com/lightningdevkit/rust-lightning/commit/57bca64bc1b7b3bbfc34fffa395cf9a38d5f8d48)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1077,35 @@ where
+		let reply_path = if create_reply_path {
+			let reply_path = self.create_blinded_path();
+			match reply_path.as_ref() {
+				Err(SendError::PathNotFound) => log_trace!(
+					logger, "Failed to create blinded_path {}", format_args!("{}", log_suffix)
+				),
```
We should log regardless of the error type. Also, include the error at the end of the log:

```rust
"Failed to create reply path {}: {:?}", log_suffix, e
```

shaavan replied:
Done! [57bca64](https://github.com/lightningdevkit/rust-lightning/commit/57bca64bc1b7b3bbfc34fffa395cf9a38d5f8d48)
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1077,35 @@ where
+			let reply_path = self.create_blinded_path();
+			match reply_path.as_ref() {
+				Err(SendError::PathNotFound) => log_trace!(
+					logger, "Failed to create blinded_path {}", format_args!("{}", log_suffix)
```
s/blinded_path/reply path

shaavan replied:
Done! [57bca64](https://github.com/lightningdevkit/rust-lightning/commit/57bca64bc1b7b3bbfc34fffa395cf9a38d5f8d48)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1077,35 @@ where
+			ResponseInstruction::WithoutReplyPath(response) => (response, false),
+			ResponseInstruction::NoResponse => return Ok(None),
+		};
+		let logger = WithContext::from(&self.logger, None, None, None);
```
Don't think we need this if we aren't adding any context.

shaavan replied:
Makes sense! Using a simple self.logger will be a better fit for this part of the code!
Addressed in [57bca64](https://github.com/lightningdevkit/rust-lightning/commit/57bca64bc1b7b3bbfc34fffa395cf9a38d5f8d48)
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1077,35 @@ where
+		);
+
+		let reply_path = if create_reply_path {
+			let reply_path = self.create_blinded_path();
+			match reply_path.as_ref() {
+				Err(SendError::PathNotFound) => log_trace!(
+					logger, "Failed to create blinded_path {}", format_args!("{}", log_suffix)
+				),
+				_ => (),
+			}
+			Some(reply_path?)
```
More idiomatic and clearer as:

```rust
match self.create_blinded_path() {
	Ok(reply_path) => Some(reply_path),
	Err(e) => {
		// log
		return Err(e);
	},
}
```

shaavan replied:
That makes the code way easier to read! :)
Thanks so much for the help!
Taken in [57bca64](https://github.com/lightningdevkit/rust-lightning/commit/57bca64bc1b7b3bbfc34fffa395cf9a38d5f8d48)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1077,35 @@ where
+			let reply_path = self.create_blinded_path();
+			match reply_path.as_ref() {
+				Err(SendError::PathNotFound) => log_trace!(
+					logger, "Failed to create blinded_path {}", format_args!("{}", log_suffix)
```
No need to use `format_args!` here because `format!` already gave you a string.

shaavan replied:
Makes sense!
Simplified in [57bca64](https://github.com/lightningdevkit/rust-lightning/commit/57bca64bc1b7b3bbfc34fffa395cf9a38d5f8d48)
Thanks!

shaavan commented:
Updated from [pr2996.07](https://github.com/shaavan/rust-lightning/commits/pr2996.07) to [pr2996.08](https://github.com/shaavan/rust-lightning/commits/pr2996.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.07..pr2996.08)):
Addressed @TheBlueMatt and @jkczyz comments

Changes:
1. Expanded docs for `respond` and `respond_with_reply_path` so that there purpose is more clearly explained.
2. Expand log message to also contain the error itself.
3. nit `blinded_path` -> `reply_path`.
4. Simplified `logger` -> `self.logger`, since there are no context to add with the error message within the function.
5. Improve the code style to be more idiomatic.
6. Simplified `format_args!` -> `format!` for `log_trace!`

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -267,14 +267,27 @@ impl Responder {
+
+	/// Creates a [`ResponseInstruction::WithReplyPath`] for a given response.
+	///
+	/// Use when the recipient need to send back a reply to us.
```
s/need/needs

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1081,36 @@ where
+		};
+
+		let message_type = response.message.msg_type();
+		let log_suffix = format!(
+			"when responding with {} to an onion message with path_id {:02x?}",
+			message_type,
+			response.path_id
+		);
```
We may actually want to inline this in both places using `format_args!`. Otherwise, we are allocating a string even if the user's log level doesn't trigger any logging.

shaavan commented:
Updated from [pr2996.08](https://github.com/shaavan/rust-lightning/commits/pr2996.08) to [pr2996.09](https://github.com/shaavan/rust-lightning/commits/pr2996.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.08..pr2996.09)):
Addressed @jkczyz comment

Update:
1. Fix grammatical error.
2. Update logging so that a string will not be created in case it is not being logged.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1081,37 @@ where
+				Err(err) => {
+					log_trace!(
+						self.logger,
+						"Failed to create reply_path when responding with {} to an onion message \
```
s/reply_path/reply path

shaavan replied:
Done! Thanks!
[27899b1](https://github.com/lightningdevkit/rust-lightning/commit/27899b1a04bfd78db34f2c22ad489f1a6fd4e5e4)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -267,14 +267,27 @@ impl Responder {
+	/// Creates a [`ResponseInstruction::WithReplyPath`] for a given response.
+	///
+	/// Use when the recipient needs to send back a reply to us.
+	pub fn respond_with_reply_path<T: OnionMessageContents>(self, response: T) -> ResponseInstruction<T> {
+		ResponseInstruction::WithReplyPath(OnionMessageResponse {
+			message: response,
+			reply_path: self.reply_path,
+			path_id: self.path_id,
+		})
+	}
```
Could you add a test for this behavior?

shaavan replied:
Introduce a new test in [27899b1](https://github.com/lightningdevkit/rust-lightning/commit/27899b1a04bfd78db34f2c22ad489f1a6fd4e5e4) to test this behavior!
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1047,18 +1081,37 @@ where
+						with path_id {:02x?}: {:?}",
+						message_type, response.path_id, err
+					);
+					return Err(err);
```
Let's add a test for the error case.

shaavan replied:
Introduce a new test in [27899b1](https://github.com/lightningdevkit/rust-lightning/commit/27899b1a04bfd78db34f2c22ad489f1a6fd4e5e4) that conditionally checks for the failure case!
Thanks for the pointer!

shaavan commented:
Updated from [pr2996.09](https://github.com/shaavan/rust-lightning/commits/pr2996.09) to [pr2996.10](https://github.com/shaavan/rust-lightning/commits/pr2996.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.09..pr2996.10)):
Addressed @jkczyz comments

Changes:
1. Corrected a nit, `reply_path` -> `reply path`
2. Expanded SendError::PathNotFound documentation as it can also be triggered when we fail to create a reply_path.
3. Introduce two new variants in TestCustomMessage, to allow an arbitrary number of back-and-forth communication, allowing test response functionality properly.
4. Introduce a new test to test the reply_path creation success, creation failure, and usage functionality.

shaavan commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -143,7 +143,7 @@ impl OffersMessageHandler for IgnoringMessageHandler {
 impl CustomOnionMessageHandler for IgnoringMessageHandler {
 	type CustomMessage = Infallible;
-	fn handle_custom_message(&self, _message: Self::CustomMessage, _responder: Option<Responder>) -> ResponseInstruction<Self::CustomMessage> {
+	fn handle_custom_message(&self, _message: Self::CustomMessage, _responder: Option<Responder>, _add_reply_path: bool) -> ResponseInstruction<Self::CustomMessage> {
```
I wanted to experiment with the `respond_with_reply_path` function and expand the capabilities of `handle_custom_message` to support creating `ResponseInstruction::WithReplyPath` through it.

However, I now see that expanding the public API should be done after a deeper discussion.

So, for the time being, I've introduced a new test-specific handler in [bb68c35](https://github.com/lightningdevkit/rust-lightning/commit/bb68c35225c80c3e900591e976d7ac3388724cf6) to facilitate both `respond` and `respond_with_reply_path` tests, while keeping the public API unchanged.

Thanks a bunch for the pointer, Matt!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -595,6 +615,9 @@ pub trait CustomOnionMessageHandler {
 	/// The returned [`Self::CustomMessage`], if any, is enqueued to be sent by [`OnionMessenger`].
 	fn handle_custom_message(&self, message: Self::CustomMessage, responder: Option<Responder>) -> ResponseInstruction<Self::CustomMessage>;
 
+	#[cfg(test)]
+	fn test_handle_custom_message(&self, message: Self::CustomMessage, responder: Option<Responder>, add_reply_path: bool) -> ResponseInstruction<Self::CustomMessage>;
+
```
You shouldn't need to expand the trait, even if it is test-only. Instead, either make a separate implementation that uses a different `ResponseInstruction` or modify the implementation to reply differently based some field, for instance. Then have the test set the field as needed.

A more general approach is to give the test implementation a set of expectations/actions (e.g., expect that the parameter to the method is X and return Y). For examples, take a look at `TestRouter`, `TestScorer`, `TestKeysInterface`, and `TestChainSource` in https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/util/test_utils.rs. They all utilize this approach in slightly different ways.

shaavan replied:
Thanks a lot for the suggestion and the references!

In [5e99d68](https://github.com/lightningdevkit/rust-lightning/commit/5e99d68c383addc464f24bd6049c8aaac73e99c2):
Instead of creating a new test-specific `handle_custom_message`, I have utilized the newly introduced `TestCustomMessage` variants to allow the `respond_with_reply_path` test.
This allowed me to leverage the `expect_message` setup too for the `TestCustomMessageHandler`.

Also, I have expanded the variants docs to explain this behavior properly.
Let me know if this approach seems sound!
Thanks a lot!

shaavan commented:
Updated from [pr2996.11](https://github.com/shaavan/rust-lightning/commits/pr2996.11) to [pr2996.12](https://github.com/shaavan/rust-lightning/commits/pr2996.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.11..pr2996.12)):
Addressed @jkczyz comment.

1. Removed the test-specific handle_custom_message, and instead utilized the newly introduced `TestCustomMessage` variants to facilitate the `respond_with_reply_path` test.
2. Expanded the `TestCustomMessage` docs, to properly explain each variant's role.

shaavan commented:
Updated from [pr2996.12](https://github.com/shaavan/rust-lightning/commits/pr2996.12) to [pr2996.13](https://github.com/shaavan/rust-lightning/commits/pr2996.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.12..pr2996.13)):
Addressed @jkczyz comment

Changes:
1. Removed redundant diff to keep commits atomic.
2. Improved the `SendError::PathNotFound` docs.
3. Removed the redundant granular comments from the added test.

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -554,7 +570,11 @@ pub enum SendError {
+	///
+	/// This occurs when:
+	/// 1. No path from sender to destination is found for sending the onion message.
+	/// 2. No reply path from destination to sender is created when responding to the onion message.
```
Done! [2b2c292](https://github.com/lightningdevkit/rust-lightning/commit/2b2c2926862eb2ed7ceba5e1e9638d287f6953ab)
Thanks!

shaavan commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -190,7 +190,6 @@ impl CustomMessageHandler for IgnoringMessageHandler {
 		// Since we always return `None` in the read the handle method should never be called.
 		unreachable!();
 	}
-
```
Cleaned up in [2b2c292](https://github.com/lightningdevkit/rust-lightning/commit/2b2c2926862eb2ed7ceba5e1e9638d287f6953ab)
Thanks!

shaavan commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -456,6 +461,67 @@ fn async_response_over_one_blinded_hop() {
+	// 4. Create a responder using the reply path for Alice.
+	let responder = Some(Responder::new(reply_path, path_id));
+
+	// 5. Expect Alice to receive the message and create a response instruction for it.
```
Made the comments less granular, and kept the comments for only the crucial steps in [2b2c292](https://github.com/lightningdevkit/rust-lightning/commit/2b2c2926862eb2ed7ceba5e1e9638d287f6953ab)
Thanks for the suggestion!

shaavan commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -80,20 +80,61 @@ impl OffersMessageHandler for TestOffersMessageHandler {
+	///
+	/// [`Request`]: TestCustomMessage::Request
 	Response,
+
+	/// A variant to enable response ping-pong.
+	/// When this variant is used, [`handle_custom_message`] interprets that
+	/// a reply path should be appended to the onion message to facilitate response
+	/// ping-pong. Consequently, it will create the corresponding
+	/// [`ResponseInstruction::WithReplyPath`] for it.
+	///
+	/// [`handle_custom_message`]: TestCustomMessageHandler::handle_custom_message
+	ResponseA,
+
+	/// The counterpart to the [`ResponseA`] variant for the response
+	/// ping-pong mechanism.
+	/// Similar to [`ResponseA`], when this variant is used, [`handle_custom_message`]
+	/// interprets that a reply path should be appended to the onion message
+	/// to facilitate response ping-pong. Therefore, it will create the corresponding
+	/// [`ResponseInstruction::WithReplyPath`] for it.
+	///
+	/// [`handle_custom_message`]: TestCustomMessageHandler::handle_custom_message
+	/// [`ResponseA`]: TestCustomMessage::ResponseA
+	ResponseB,
 }
```
Thanks a lot for the suggestion!

I can definitely see that the approach you described will be a cleaner solution for the `respond_with_reply_paths` tests.

That said let me share my thought process behind opting for a different set of variants initially. I  wanted to distinguish clearly between the “one request - one response flow” and the “ping-pong flow”.

For me, part of it came down to the connotations of the words 'Request' and 'Response'. Sending a 'Request' as a response felt a bit… awkward…? So, I ended up leaning towards those new variants to streamline the ping-pong interactions.

Your approach does seem like a better way to handle the tests without introducing extra variants. I'm just wondering if we should also consider renaming 'Request' and 'Response' to better suit their general purpose.

I'd love to hear your thoughts on this! Your suggestions are always super valuable.

Thanks a lot, once again, Jeff!

jkczyz replied:
Renaming the variants to `Ping` and `Pong` is fine by me. Please do so in a separate commit.

shaavan replied:
Done! 🚀
Renamed the `Request` and `Response` to `Ping` and `Pong`, and introduced the `include_reply_path` function in [f5daf5f](https://github.com/lightningdevkit/rust-lightning/commit/f5daf5f8c80b749bfed0eaabfebab2e0e3207ffc)
Thanks a lot for the guide points! :)

shaavan commented:
Updated from [pr2996.13](https://github.com/shaavan/rust-lightning/commits/pr2996.13) to [pr2996.14](https://github.com/shaavan/rust-lightning/commits/pr2996.14) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.13..pr2996.14)):
Addressed @jkczyz comment

Changes:
1. Renamed `Request` and `Response` to `Ping` and `Pong` to better suit their new roles of being each other's response.
2. Removed the now redundant variants `ResponseA` and `ResponseB`.
3. Introduced a new variable in `TestCustomMessageHandler` that signals to `handle_custom_message` which `ResponseInstruction` it should create.
4. Updated the introduced test to utilize this new setup.

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -112,16 +112,31 @@ impl Writeable for TestCustomMessage {
 
 struct TestCustomMessageHandler {
 	expected_messages: Mutex<VecDeque<TestCustomMessage>>,
```
I wonder if we should just make this hold a `(TestCustomMessage, bool)` and add an `expect_message_and_respond` method instead of adding `include_reply_path`. Then you don't need to reset the `bool` after handling a message.

shaavan replied:
Thanks a lot for the idea!
Built further upon the idea after offline discussion.

1. Structured the expectations in the `Expectation` struct.
2. Introduce the `expect_message_and_response` function in [594dc0f](https://github.com/lightningdevkit/rust-lightning/commit/594dc0fd893eb85abde3ee7cc8527b225b042cb4)

shaavan commented:
Updated from [pr2996.14](https://github.com/shaavan/rust-lightning/commits/pr2996.14) to [pr2996.15](https://github.com/shaavan/rust-lightning/commits/pr2996.15) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.14..pr2996.15)):
Addressed @jkczyz comment

Changes:
1. Introduce the add_reply_path boolean as part of expectations. This allows saving up on a mutex variable and keeps things consistent with the current test suite.
2. Introduce a new struct `Expectation` to keep the current (and future) expectations neatly ordered.

shaavan commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -112,16 +112,31 @@ impl Writeable for TestCustomMessage {
 
 struct TestCustomMessageHandler {
 	expected_messages: Mutex<VecDeque<TestCustomMessage>>,
+	add_reply_path: Mutex<bool>,
```
Gone with the `Expectation` struct approach in [594dc0f](https://github.com/lightningdevkit/rust-lightning/commit/594dc0fd893eb85abde3ee7cc8527b225b042cb4)
Thanks!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -401,6 +429,62 @@ fn async_response_over_one_blinded_hop() {
+}
+
+#[test]
+fn async_response_with_reply_path_over_one_blinded_hop() {
+	do_test_async_response_with_reply_path_over_one_blinded_hop(true);
+	do_test_async_response_with_reply_path_over_one_blinded_hop(false);
+}
```
IMO, you can drop `_over_one_blinded_hop` as it isn't pertinent to what is being tested.

shaavan replied:
I think this will definitely be a naming improvement!
Addressed in [e6effd4](https://github.com/lightningdevkit/rust-lightning/commit/e6effd447d47e1c3c7771935bea58801199b60b3)

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -401,6 +429,62 @@ fn async_response_over_one_blinded_hop() {
+	// Alice handles the message reponse, and creates the appropriate ResponseInstruction for it.
+	let response_instruction = alice.custom_message_handler.handle_custom_message(message, responder);
+
+	if !reply_path_succeed {
```
I'd rather have two separate tests than use the utility function taking `reply_path_succeed`. They would be much easier to read without the conditional logic, and the setup code is trivial.

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -401,6 +429,62 @@ fn async_response_over_one_blinded_hop() {
+	let bob = &nodes[1];
+
+	// Alice receives a message from Bob with a reply path
+	let message = TestCustomMessage::Ping;
```
The test would be more readable if this were inlined where needed.

shaavan replied:
I have updated the test flow and structure for better readability.
With this new structure, I have kept the `message` variable to make it easier to understand the flow.
Let me know if the updated tests look good!
Thanks!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -401,6 +429,62 @@ fn async_response_over_one_blinded_hop() {
+
+	// Alice receives a message from Bob with a reply path
+	let message = TestCustomMessage::Ping;
+	let path_id = Some([2; 32]);
```
Could you move this to where it is used?

shaavan replied:
I have moved the message, path_id, and reply_path (sent by Bob to us) together to update the test readability.
Let me know if the updated test flow looks good!
Thanks!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -111,16 +111,37 @@ impl Writeable for TestCustomMessage {
 
 struct TestCustomMessageHandler {
-	expected_messages: Mutex<VecDeque<TestCustomMessage>>,
+	expected_messages: Mutex<VecDeque<Expectations>>,
```
Let's rename the field `expectations`

shaavan replied:
Done! [e6effd4](https://github.com/lightningdevkit/rust-lightning/commit/e6effd447d47e1c3c7771935bea58801199b60b3)
Thanks!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -111,16 +111,37 @@ impl Writeable for TestCustomMessage {
+	expected_messages: Mutex<VecDeque<Expectations>>,
+}
+
+struct Expectations {
+	message: TestCustomMessage,
+	add_reply_path: bool,
 }
```
How about this?

```rust
struct OnHandleCustomMessage {
	expect: TestCustomMessage,
	respond_with: Option<TestCustomMessage>,
	include_reply_path: bool,
}
```

Probably don't need `respond_with` now, but just wanted to show how this could be expanded in the future if needed.

shaavan replied:
I think the naming sounds better, as it makes it more expressive in what it does.
Took up your suggestion in [e6effd4](https://github.com/lightningdevkit/rust-lightning/commit/e6effd447d47e1c3c7771935bea58801199b60b3)
Thanks!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -138,16 +159,23 @@ impl Drop for TestCustomMessageHandler {
-			Some(expected_msg) => assert_eq!(expected_msg, msg),
+		let add_reply_path = match self.expected_messages.lock().unwrap().pop_front() {
+			Some(Expectations { message: expected_msg, add_reply_path }) => {
+				assert_eq!(expected_msg, msg);
```
nit: put the expectation second

shaavan replied:
Done! [e6effd4](https://github.com/lightningdevkit/rust-lightning/commit/e6effd447d47e1c3c7771935bea58801199b60b3)

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -138,16 +159,23 @@ impl Drop for TestCustomMessageHandler {
+			}
 		} else {
 			ResponseInstruction::NoResponse
 		}
```
If the response can no longer be `None`, then you don't need to use an `Option`.

shaavan replied:
Makes sense!
Refactored the code in [e6effd4](https://github.com/lightningdevkit/rust-lightning/commit/e6effd447d47e1c3c7771935bea58801199b60b3)
Thanks!

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -138,16 +159,23 @@ impl Drop for TestCustomMessageHandler {
+				responder.respond(response)
+			}
 		} else {
 			ResponseInstruction::NoResponse
```
If the expectation specified a reply path should be added, we should probably panic if we don't have a responder to send back a response.

shaavan replied:
Makes sense!
I think introducing an explicit panic will allow tests to be more specific.
Introduced it in [e6effd4](https://github.com/lightningdevkit/rust-lightning/commit/e6effd447d47e1c3c7771935bea58801199b60b3)
Thanks!

shaavan commented:
Updated from [pr2996.15](https://github.com/shaavan/rust-lightning/commits/pr2996.15) to [pr2996.16](https://github.com/shaavan/rust-lightning/commits/pr2996.16) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.15..pr2996.16)):
Addressed @jkczyz comments

1. Refactor the function and improve naming structs for TestCustomMessageHandler.
2. Improve the test readability by properly structuring and separating success and failure cases in separate tests.

shaavan commented:
Updated from [pr2996.16](https://github.com/shaavan/rust-lightning/commits/pr2996.16) to [pr2996.17](https://github.com/shaavan/rust-lightning/commits/pr2996.17) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.16..pr2996.17)):

Changes:
1. Rebase on main

shaavan commented:
@jkczyz, @TheBlueMatt 
A gentle ping! :)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1043,20 +1081,41 @@ where
 	}
 
-	fn handle_onion_message_response<T: OnionMessageContents>(
+	/// Handles the response to an [`OnionMessage`] based on its [`ResponseInstruction`],
+	/// enqueueing any response for sending.
+	pub fn handle_onion_message_response<T: OnionMessageContents>(
```
Let's add further information on how to use this in a second paragraph. Something like:

```rust
/// Useful for handling [`OnionMessage`]s asynchronously. A handler may return
/// [`ResponseInstruction::NoResponse`] and move the passed [`Responder`] to another task for
/// generating a response. That task can then call this method once ready to send the response.
```

shaavan replied:
Sounds good!
Expanded the docs in [834779d](https://github.com/lightningdevkit/rust-lightning/commit/834779de16284f0da3c47b42ca9ac6cd004a142e)
Let me know if it looks good!
Thanks!

jkczyz reviewed (commented):
LGTM other than expanding the docs a bit.

shaavan commented:
Updated from [pr2996.17](https://github.com/shaavan/rust-lightning/commits/pr2996.17) to [pr2996.18](https://github.com/shaavan/rust-lightning/commits/pr2996.18) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.17..pr2996.18)):
Addressed @jkczyz comment

Update:
1. Rebase on main.
2. Update the `create_blinded_path` function and `BlindedPath::new_for_message` to match the API change.
3. Expanded `handle_onion_message_response` docs.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -981,6 +1001,33 @@ where
+			.map_err(|_| SendError::GetNodeIdFailed)?;
+		let secp_ctx = &self.secp_ctx;
+
+		// let peers = self.message_recipients.lock().unwrap()
+		// 	.iter()
+		// 	.filter(|(_, peer)| matches!(peer, OnionMessageRecipient::ConnectedPeer(_)))
+		// 	.map(|(node_id, _)| *node_id)
+		// 	.collect();
```
Looks like this should be removed.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1053,20 +1100,49 @@ where
+	/// This function is useful for asynchronous handling of [`OnionMessage`]s.
+	///
+	/// Handlers have the option to return [`ResponseInstruction::NoResponse`], indicating that
+	/// no immediate response should be sent. Instead, they can transfer the associated [`Responder`]
```
"Instead" indicates an alternative to what was just stated. So use "Then" instead (no pun intended).

shaavan replied:
I think that makes more sense.
Used "then" instead of "instead" in [d8adbb0](https://github.com/lightningdevkit/rust-lightning/commit/d8adbb0a5b7dcc39a41596274129ff3d69ec08de) (pun intended :)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1053,20 +1100,49 @@ where
+	/// Handles the response to an [`OnionMessage`] based on its [`ResponseInstruction`],
+	/// enqueueing any response for sending.
+	///
+	/// This function is useful for asynchronous handling of [`OnionMessage`]s.
+	///
+	/// Handlers have the option to return [`ResponseInstruction::NoResponse`], indicating that
+	/// no immediate response should be sent. Instead, they can transfer the associated [`Responder`]
+	/// to another task responsible for generating the response asynchronously. Subsequently, when
+	/// the response is prepared and ready for sending, that task can invoke this method to enqueue
+	/// the response for delivery.
```
Let's combine these paragraphs as the second is describing asynchronous handling.

shaavan replied:
Done! [d8adbb0](https://github.com/lightningdevkit/rust-lightning/commit/d8adbb0a5b7dcc39a41596274129ff3d69ec08de)
Thanks!

jkczyz reviewed (commented):
BTW, please avoid rebasing mid-review as it makes it difficult for reviewers to tell what has changed.

shaavan commented:
Updated from [pr2996.18](https://github.com/shaavan/rust-lightning/commits/pr2996.18) to [pr2996.19](https://github.com/shaavan/rust-lightning/commits/pr2996.19) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2996.18..pr2996.19)):
Addressed @jkczyz's comment:

Changes:
1. Removed the redundant commented code.
2. Applied nit fixes to docs.

@jkczyz:
> BTW, please avoid rebasing mid-review as it makes it difficult for reviewers to tell what has changed.

Understood! I tried to keep the changes with rebase separate, but for this one, I took a shortcut since the suggestion was a small one. I can see how that would have made reviewing difficult. I’ll make sure to keep the rebase and changes separate in the future. Thanks!

shaavan commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -132,25 +155,32 @@ impl Drop for TestCustomMessageHandler {
+
+		let response = match msg {
+			TestCustomMessage::Ping => TestCustomMessage::Pong,
+			TestCustomMessage::Pong => TestCustomMessage::Ping,
```
The new test for this PR required multiple back-and-forth communication between the nodes. To allow doing so we settled on making Ping and Pong each other's response.

Here's a [discussion thread](https://github.com/lightningdevkit/rust-lightning/pull/2996#discussion_r1608660306) related to it!

Thanks!
